pkg/kp/rollstore/consul_store.go

```diff
@@ -0,0 +1,555 @@
+	var newRCID rc_fields.ID
+	select {
+	case err = <-renewalErrCh:
+		return roll_fields.Update{}, err
+	default:
+		rc, err := s.rcstore.Create(newRCManifest, newRCNodeSelector, newRCPodLabels)
+		if err != nil {
+			return roll_fields.Update{}, err
+		}
+
+		newRCCleanup = func() {
+			err := s.rcstore.Delete(newRCID, false)
+			if err != nil {
+				s.logger.WithError(err).Errorln("Unable to cleanup RC %s after failed RU creation attempt", newRCID)
+			}
+		}
+
+		newRCID = rc.ID
+
+		// Get a lock on the new RC we just created so no parallel
+		// update creations can use it
+		err = s.lockRCs(rc_fields.IDs{newRCID}, session)
+		if err != nil {
+			return roll_fields.Update{}, err
+		}
+	}
+
+	rcIDs = append(rcIDs, newRCID)
+	// Check for conflicts again in case an update was created on the new
+	// RC between when we created it and locked it
+	err = s.checkForConflictingUpdates(rcIDs)
+	if err != nil {
+		return roll_fields.Update{}, err
+	}
+
+	u = roll_fields.Update{
+		OldRC:           oldRCID,
+		NewRC:           newRCID,
+		DesiredReplicas: desiredReplicas,
+		MinimumReplicas: minimumReplicas,
+		LeaveOld:        leaveOld,
+		RollDelay:       rollDelay,
+	}
+
+	return s.attemptRUCreation(u, renewalErrCh)
+}
+
+// Creates a rolling update that may or may not already have an existing old
+// RC. If one matches the oldRCSelector, it will be used as the old RC in the
+// new update.  If one does not exist, a "dummy" old RC will be created that is
+// identical to the specifications for the new RC.
+// Returns an error if the old RC exists but is part of another RU, or if
+// the label selector returns more than one match.
+func (s consulStore) CreateRollingUpdateFromOneMaybeExistingWithLabelSelector(
+	oldRCSelector klabels.Selector,
+	desiredReplicas int,
+	minimumReplicas int,
+	leaveOld bool,
+	rollDelay time.Duration,
+	newRCManifest pods.Manifest,
+	newRCNodeSelector klabels.Selector,
+	newRCPodLabels klabels.Set,
+) (u roll_fields.Update, err error) {
+	// This function may or may not create old and new RCs and subsequently
+	// fail, so we defer a function that does any cleanup (if applicable)
+	var cleanupOldRC func()
+	var cleanupNewRC func()
+
+	defer func() {
+		if err != nil {
+			if cleanupOldRC != nil {
+				cleanupOldRC()
+			}
+
+			if cleanupNewRC != nil {
+				cleanupNewRC()
+			}
+		}
+	}()
+
+	session, renewalErrCh, err := s.newRUCreationSession()
+	if err != nil {
+		return roll_fields.Update{}, err
+	}
+
+	// Check if any RCs match the oldRCSelector
+	matches, err := s.labeler.GetMatches(oldRCSelector, labels.RC)
+	if err != nil {
+		return roll_fields.Update{}, err
+	}
+
+	if len(matches) > 1 {
+		return roll_fields.Update{}, util.Errorf(
+			"Can't create update: old RC selector %s was ambiguous, provided %d matches",
+			oldRCSelector.String(),
+			len(matches),
+		)
+	}
+
+	var oldRCID rc_fields.ID
+	if len(matches) == 1 {
+		oldRCID = rc_fields.ID(matches[0].ID)
+	} else {
```

Comment:
it's above up above, but probably putting the whole chain together would make that clearer